doc/xftlsfonts.1
.TH xftlsfonts 1 "August 2002" "xftlsfonts" "v1.0"
.SH NAME
xftlsfonts - display Xft fonts found
.SH SYNOPSIS
\fBxftlsfonts\fR [options]
.SH DESCRIPTION
\fBxftlsfonts\fR is intended to display the Xft fonts that the X Window System
currently supports.
.SH OPTIONS
.TP
\fB\-style\fR
Show possible styles for each of the fonts.
.TP
\fB\-slant\fR
Show the slant for each of the fonts.
.TP
\fB\-weight\fR
Show the weight for each of the fonts.
.TP
\fB\-file\fR
Show which file contains each of the fonts.
.TP
.SH CONFIGURATION
\fIxftlsfonts\fR will work without any arguments, assuming a default behaviour.
.SH AUTHOR
The author of xftlsfonts is Ben Jansens <\fIben@orodu.net\fB>.
.SH SEE ALSO
\fIopenbox\fR(1)


CHANGELOG
3.7:
  * Update translations.
  * Fix ShapeInput shape mask propagation (clickthrough windows).
  * Add fullscreen, focusable conditions to If and ForEach actions.
  * Add <none> to ForEach, ran if there were no matches.
    (The actions in the <else> tag are run for every checked client.)
  * Allow ignoring repeated keyboard events with XKB, using the repeat="no"
    attribute. Useful to not accidentally close every window if your F4 key
    gets stuck.
  * Fix a crash that was exposed by glib removing their arena allocator.
  * Add the FocusFallback action which does the same as the Unfocus action
    but with a more logical name. The Unfocus action may change to unfocus the
    window in a future release (and not focus any other window in its place).
  * Various small bugfixes.
  * Added focus.cycle.bg.color and focus.cycle.border.color theme options.

3.6.1:
  * Bump ABI version of libobrender that should have been done in 3.6

3.6:
  * Added various new abilities to If action, including the <query> tag,
    matching class, name, role and window type.
  * Added ForEach action.
  * Restore the <center> option for window placement.
  * New GrowToFill action.
  * Updated Portuguese and Polish translations.
  * New Galician translation.
  * Added strict option to ToggleShowDesktop action.
  * Speed optimization in rendering code.
  * New <position> tag for ShowMenu action allows placing menus other
    than at the mouse cursor.
  * Fix some corner cases of relative position to work correctly.
  * Allow specifying multiple mousebinds in one tag as we already do
    for keybinds (separated by spaces, as "W-Return M-S-F3" for example).
  * Added option rebindOnMappingNotify option, on by default, which
    controls if we update binds when the keyboard layout changes at runtime.
  * The resize action now only considers the part of the client that is onscreen
    for purposes of picking an action to perform.
  * Make desktop actions non-interactive (means you can have more actions bound
    to a key after them, and the dialog disappears on a timer instead of when
    releasing the modifier keys).
  * obxprop is now not pathologically slow when printing very long arrays.
  * Fix menu accelerator underlines sometimes not showing up.
  * Fix so you can click menu entries before the menuHideDelay expires if it's
    a new click.
  * Don't readd the X event fd to the poll set every time we reconfigure.
  * Fix some specialized button bitmaps not loading from the correct files.
    (toggled + pressed or hover would just be the base toggled image).
  * Various small bugfixes.

3.5.2:
  * Fix crash in theme rendering code, and in theme parsing code.
  * Maximize windows without borders to fill the whole screen, don't leave
    a one-pixel empty space. (Bug #5996)

3.5.1:
  * New translations: Afrikaans, Belarusian, Interlingua, Hebrew, Romanian,
    Greek.
  * Updated translations: Italian, Serbian, Spanish, Czech, Hungarian, Turkish,
    German, Arabic, Polish, Dutch, Lithuanian, Portuguese, Estonian.
  * SVG icon support. This is optional, depending on librsvg being installed.
  * Allow application rules to control window size with a new <size> tag.
  * Allow application rules to pick a monitor for new windows without forcing
    a position.
  * Allow non-interactive focus cycling with a new <interactive> tag.
  * New LeastOverlap window placement policy replaces the old default
    behaviour. It finds a place on a given monitor that overlaps as few
    windows as possible. Contributed by Ian Zimmerman.
  * Improved Xinerama behaviour.
  * Correctly interface with latest gnome-session for Gnome/Openbox X sessions.
  * Allow third-party control of window opacity in compositing managers.
  * Improved themeing options. Contributed by Dave Foster.
  * Add <monitor>, <title type="regex">, <title type="exact"> and
    <activedesktop> options to If action.
  * Addresses bugs #4661, #5506, #5186, #5758, #5410, #5228, #5277, #5731,
    #5746, #5737, #5419, #5721, #5711, #5385, #5500, #4992, #5443, #5518,
    #5444, #4782, #5237, #5228, #5173, #5203, #5246, #5180, #5179, #5150,
    #5132, #4937, #4889, #5253, #3769, #5819, #5811, #5081, #5426 among others.

3.5.0:
  * New alt-tab dialog shows windows in a vertical list.
  * Improved Xinerama support.
  * Allow icons in menus.
  * Theme options for prompt dialogs (osd.button.unpressed.*,
    osd.button.pressed.*, osd.button.focused.*)
  * Addresses bug #4877, #4596, #4617, #4752, #4663, #4662, #4586, #2319,
    #4341, #4519, #4543, #4503, #4355, #4072, #3702, #4284
  * Lots of additional bug fixes and performance improvements.

3.4.11.2:
  * Updated Estonian and Portuguese translations.
  * Fix a rare crash involving moving fullscreen windows to different monitors
  * Fix a more common crash involving pressing right in a menu

3.4.11.1:
  * Updated Polish translation.
  * Fixed bug #4519 (Incorrect focus in reused windows).
  * Lower the default submenu open/hide delay from 200ms to 100ms.
  * Fix some more problems with gnome integration.
  * Update Clearlooks theme.
  * Some other small fixes.

3.4.11:
  * Update Hungarian, Japanese, and Latvian translations.
  * Make xdg-autostart use the OPENBOX environment by default, so you can use
    OnlyShowIn=OPENBOX in an autostart .desktop and it will work as expected.
  * Don't close the menu when you hold control and execute something.
  * Fix bug #4503 (Adjust who shows up in the Alt-Tab list using SKIP_TASKBAR).
  * Fix flickering window when moving maximized window between monitors of
    different sizes.
  * Fix bug #4355 (Allow multiple escaped _'s in a menu label and allow
    a menu shortcut to come after an escaped _).
  * Remember the maximized state of a window when it goes fullscreen, and
    restore it when leaving fullscreen state.
  * Fix bug #4072 (Openbox is stopped by terminal applications writing to
    stdout).
  * Fix bug #4492 (Mistake in openbox-gnome-session check while setting up).

util/xftlsfonts.cc
const char *NAME = "xftlsfonts";
const char *VERSION = "1.0";

#ifdef    HAVE_CONFIG_H
#include "../config.h"
#endif // HAVE_CONFIG_H

#ifndef   XFT
#include <iostream>
using std::cout;
using std::endl;

int main(int, char **) {
  cout << NAME << " version " << VERSION << endl;
  cout << "Copyright (c) 2002, Ben Jansens <ben@orodu.net>" << endl;
  cout << endl;
  cout << "Openbox was built without support for Xft fonts. This utility must"
    << endl;
  cout << "must be built with Xft support in order to function." << endl;
}
#else  // XFT
extern "C" {
#include <X11/Xlib.h>
#include <X11/Xft/Xft.h>
}

#include <iostream>
#include <string>
#include <vector>

using std::string;
using std::cout;
using std::endl;

int main(int argc, char **argv) {
  if (argc > 1)
    for (int i = 1; i < argc; ++i)
      if (string(argv[i]) == "-help" ||
          string(argv[i]) == "--help" ||
          string(argv[i]) == "-version" ||
          string(argv[i]) == "--version") {
        cout << NAME << " version " << VERSION << endl;
        cout << "Copyright (c) 2002, Ben Jansens <ben@orodu.net>" << endl;
        cout << endl;
        cout << "Usage: " << argv[0] << " [options]" << endl;
        cout << "    -style     Show possible styles for each font" << endl;
        cout << "    -slant     Show the slant for each font" << endl;
        cout << "    -weight    Show the weight for each font" << endl;
        cout << "    -file      Show which files contain each font" << endl;
        cout << endl;
        return 1;
      }

  Display *display = XOpenDisplay(NULL);
  if (! display) {
    cout << "Failed to open connection to X display\n";
    return 2;
  }

  XftObjectSet *obj = XftObjectSetCreate();
  if (! obj) {
    cout << "Failed to create an XftObjectSet\n";
    return 2;
  }

  XftObjectSetAdd(obj, XFT_FAMILY);

  if (argc > 1)
    for (int i = 1; i < argc; ++i) {
      if (string(argv[i]) == "-style") XftObjectSetAdd(obj, XFT_STYLE);
      else if (string(argv[i]) == "-file") XftObjectSetAdd(obj, XFT_FILE);
      else if (string(argv[i]) == "-slant") XftObjectSetAdd(obj, XFT_SLANT);
      else if (string(argv[i]) == "-weight") XftObjectSetAdd(obj, XFT_WEIGHT);
    }

  XftPattern *pat = XftPatternCreate();
  if (! pat) {
    cout << "Failed to create an XftPattern\n";
    exit(2);
  }

  XftFontSet *set = XftListFontsPatternObjects(display, DefaultScreen(display),
                                               pat, obj);
  if (! set) {
    cout << "Failed to find a matching XftFontSet\n";
    exit(2);
  }
 
  XFree(pat);
  XFree(obj);

  for (int i = 0; i < set->nfont; ++i) {
    for (int e = 0; e < set->fonts[i]->num; ++e) {
//      if (string(set->fonts[i]->elts[e].object) != "family")
//        continue; // i just want font family names

      if (e > 0)
        cout << "  "; // indent after the first element
      cout << set->fonts[i]->elts[e].object << ": ";

      XftValueList *vallist = set->fonts[i]->elts[e].values;
      bool f = true;
      do {
        if (f)
          f = false;
        else
          cout << ", ";

        XftValue val = vallist->value;
        switch (val.type) {
        case XftTypeVoid:
          cout << "(void)";
          break;

        case XftTypeInteger:
          cout << val.u.i;
          break;

        case XftTypeDouble:
          cout << val.u.d;
          break;

        case XftTypeString:
          cout << val.u.s;
          break;

        case XftTypeBool:
          cout << val.u.b;
          break;

        case XftTypeMatrix:
          cout << "xx(" << val.u.m->xx << ") ";
          cout << "xy(" << val.u.m->xy << ") ";
          cout << "yx(" << val.u.m->yx << ") ";
          cout << "yy(" << val.u.m->yy << ")";
          break;
        }
      } while ((vallist = vallist->next));
      cout << endl;
    }
  }
  
  cout << endl << "Found " << set->nfont << " matches." << endl;

  XFree(set);
  
  XCloseDisplay(display);
  return 0;
}
#endif // XFT

